File: cowebx-apps/comap/src/main/java/org/coweb/example/ZipVisits.java
/**
 * Copyright (c) The Dojo Foundation 2011. All Rights Reserved.
 * Copyright (c) IBM Corporation 2008, 2011. All Rights Reserved.
 */
package org.coweb.example;

import org.coweb.bots.VanillaBot;
import org.coweb.bots.Proxy;

import java.util.Timer;
import java.util.Map;
import java.util.HashMap;
import java.util.TimerTask;
import java.util.Random;

/**
 * Simple bot that listens for messages from the moderator and sends out
 * artificially generated visitor counts every five seconds.
 */
public class ZipVisits extends VanillaBot {

    private Proxy proxy = null;
    private int total = 0;
    private Timer timer = null;
    private Map<String, Object> visits = new HashMap<String, Object>();
    private Random r = new Random();

    /* This is necessary for all bots. The proxy is necessary for
     * communication. */
    public void setProxy(Proxy proxy) {
       this.proxy = proxy;
    }

    @Override
	public void onShutdown() {
        if (null != this.timer) {
            this.timer.cancel();
            this.timer = null;
        }
	}

    /*
     * Called upon session startup.
     */
    @Override
	public void init() {
        this.visits.clear();
        this.timer = new Timer();
        this.timer.scheduleAtFixedRate(new BotTimer(), 0, 5000);
	}

    /*
     * Handle messages from the moderator. The moderator will send us marker
     * positions via this mechanism.
     */
	public synchronized void onRequest(Map<String, Object> params, String replyToken,
            String username) {

        String uuid = (String)params.get("uuid");

        int visitCount = this.getVisitCount(params);
        this.visits.put(uuid, visitCount);

        /* The reply token is used to uniquely identify which client sent the
         * message and to distinguish between multiple messages if the client
         * has sent more than one to this service. It must be considered
         * opaque and not be altered. */
        Map<String, Object> reply = new HashMap<String, Object>();
        reply.put("reply", "acknowledged");
        this.proxy.reply(this, replyToken, reply);

	}

    private int getVisitCount(Map<String, Object> params) {
        /* params contains the latitude and longitude of the pin drop
         * location, but the truth is that we artificially generate the visit
         * count, so we don't really look at the lat/lon location. */
        synchronized (this.r) {
            return r.nextInt(1000);
        }
    }

    private void updateVisits() {
        /* We artificially increment the visit count, as if over time,
         * more visitors visit the location in question. */
        for (String uuid: this.visits.keySet()) {
            int cnt = (Integer)this.visits.get(uuid);
            synchronized (this.r) {
                cnt += this.r.nextInt(10);
            }
            this.visits.put(uuid, cnt);
        }
    }

    private class BotTimer extends TimerTask {
        @Override
        public void run() {
            ZipVisits bot = ZipVisits.this;
            bot.updateVisits();
            System.out.println("Bot: zips=" + bot.visits);
            bot.proxy.publish(bot, bot.visits);
        }
    }

}

File: cowebx-apps/comap/src/main/python/bots/zipvisits/zipvisits.py
'''
Copyright (c) The Dojo Foundation 2011. All Rights Reserved.
Copyright (c) IBM Corporation 2008, 2011. All Rights Reserved.
'''
# std lib
import random
# coweb
import coweb.bot

class ZipVisitsBot(object):
    def __init__(self, botWrapper, *args):
      self.bot = botWrapper
      self.visits = {}
      self.timer = None

    def on_request(self, params, token, username):
        uuid = params.get("uuid")
        visitCount = self.getVisitCount(params)
        self.visits[uuid] = visitCount
        self.bot.reply(token, {"reply": "acknowledged"})
        if self.timer is None:
            # start a timer to publish data every 5 sec
            self.timer = self.bot.add_timer(5, self.on_timer)

    def getVisitCount(esfl, params):
        return random.randint(0, 1000)

    def updateVisits(self):
        for uuid in self.visits:
            cnt = self.visits.get(uuid)
            self.visits[uuid] = self.visits[uuid] + random.randint(0, 10)

    def on_timer(self):
        # update counts for marker; if we weren't faking data, this is where
        # we'd go off a fetch the info
        self.updateVisits()
        # publish the updated results back to clients; include all the markers
        # but we could send just those that updated
        self.bot.publish(self.visits)
        # schedule next timer interval
        self.timer = self.bot.add_timer(5, self.on_timer)

    def on_shutdown(self):
        # make sure we're not hanging on a timer
        if self.timer:
            self.bot.remove_timer(self.timer)
            self.timer = None

coweb.bot.run(ZipVisitsBot)

